File: sdk/go/server/leaderboards.go
package server

import (
	"context"

	serverv1 "github.com/torchwoodcloud/torchwood/genproto/server/v1"
	sharedv1 "github.com/torchwoodcloud/torchwood/genproto/shared/v1"
)

// LeaderboardsService 封装 Server API 的排行榜面（代任意 subject 提交 +
// 快照/榜读 + board 配置管控）。提交窗口 = {当前期, 上一期}，封榜后拒收；
// board 管控写动词需 leaderboards.admin（与 leaderboards.write 刻意分离）。
type LeaderboardsService struct {
	c   *Client
	api serverv1.LeaderboardsServiceClient
}

func (s *LeaderboardsService) SubmitLeaderboardScore(ctx context.Context, req *serverv1.SubmitLeaderboardScoreRequest) (*sharedv1.LeaderboardScoreSnapshot, error) {
	return s.api.SubmitLeaderboardScore(ctx, req)
}

func (s *LeaderboardsService) GetLeaderboardEntry(ctx context.Context, boardID, subjectID, period string) (*sharedv1.LeaderboardScoreSnapshot, error) {
	return s.api.GetLeaderboardEntry(ctx, &serverv1.GetLeaderboardEntryRequest{BoardId: boardID, SubjectId: subjectID, Period: period})
}

func (s *LeaderboardsService) ListLeaderboardTop(ctx context.Context, req *serverv1.ListLeaderboardTopRequest) (*sharedv1.ListLeaderboardTopResponse, error) {
	return s.api.ListLeaderboardTop(ctx, req)
}

func (s *LeaderboardsService) GetLeaderboardSettlement(ctx context.Context, boardID, period string) (*serverv1.GetLeaderboardSettlementResponse, error) {
	return s.api.GetLeaderboardSettlement(ctx, &serverv1.GetLeaderboardSettlementRequest{BoardId: boardID, Period: period})
}

func (s *LeaderboardsService) ListLeaderboardSettlements(ctx context.Context, req *serverv1.ListLeaderboardSettlementsRequest) (*serverv1.ListLeaderboardSettlementsResponse, error) {
	return s.api.ListLeaderboardSettlements(ctx, req)
}

// —— board 配置管控（预置与门禁自动化；Create/Update 需 leaderboards.admin）——

// CreateLeaderboardBoard 幂等建榜：已存在且配置逐字段相等（缺省归一后）→
// 200 + 现状（重放安全）；不等 → ALREADY_EXISTS（错误消息附字段 diff）。
func (s *LeaderboardsService) CreateLeaderboardBoard(ctx context.Context, req *serverv1.CreateLeaderboardBoardRequest) (*sharedv1.LeaderboardBoard, error) {
	return s.api.CreateLeaderboardBoard(ctx, req)
}

func (s *LeaderboardsService) GetLeaderboardBoard(ctx context.Context, boardID string) (*sharedv1.LeaderboardBoard, error) {
	return s.api.GetLeaderboardBoard(ctx, &serverv1.GetLeaderboardBoardRequest{BoardId: boardID})
}

// ListLeaderboardBoards 整列返回（不分页，每项目 ≤100）。
func (s *LeaderboardsService) ListLeaderboardBoards(ctx context.Context) (*serverv1.ListLeaderboardBoardsResponse, error) {
	return s.api.ListLeaderboardBoards(ctx, &serverv1.ListLeaderboardBoardsRequest{})
}

func (s *LeaderboardsService) ListLeaderboardBoardPeriods(ctx context.Context, boardID string, limit int32) (*serverv1.ListLeaderboardBoardPeriodsResponse, error) {
	return s.api.ListLeaderboardBoardPeriods(ctx, &serverv1.ListLeaderboardBoardPeriodsRequest{BoardId: boardID, Limit: limit})
}

// UpdateLeaderboardBoard 改榜（proto3 optional 语义：未设置 = 不修改）。
func (s *LeaderboardsService) UpdateLeaderboardBoard(ctx context.Context, req *serverv1.UpdateLeaderboardBoardRequest) (*sharedv1.LeaderboardBoard, error) {
	return s.api.UpdateLeaderboardBoard(ctx, req)
}

File: sdk/typescript/src/server/leaderboards.ts
import { listQuery, type HttpTransport } from "../http.js";
import type {
  CreateLeaderboardBoardInput,
  LeaderboardBoard,
  LeaderboardScoreSnapshot,
  LeaderboardSettlement,
  LeaderboardSettlementGrant,
  ListLeaderboardTopResponse,
  SubmitLeaderboardScoreInput,
  UpdateLeaderboardBoardInput,
} from "../types.js";

/** Server API 排行榜面：代任意 subject 提交（leaderboards.write）+ 快照/榜读
 * （leaderboards.read）+ board 配置管控（leaderboards.admin，配置面与提交面
 * 刻意分离——提交分值的密钥不得改榜配置）。 */
export class ServerLeaderboardsService {
  constructor(private readonly http: HttpTransport) {}

  async submitLeaderboardScore(
    boardId: string,
    subjectId: string,
    input: SubmitLeaderboardScoreInput,
  ): Promise<LeaderboardScoreSnapshot> {
    return this.http.request<LeaderboardScoreSnapshot>(
      "POST",
      `/v1/server/leaderboards/${encodeURIComponent(boardId)}:submit`,
      { auth: "apiKey", body: { subject_id: subjectId, ...input } },
    );
  }

  async getLeaderboardEntry(
    boardId: string,
    subjectId: string,
    period?: string,
  ): Promise<LeaderboardScoreSnapshot> {
    return this.http.request<LeaderboardScoreSnapshot>(
      "GET",
      `/v1/server/leaderboards/${encodeURIComponent(boardId)}/entries/${encodeURIComponent(subjectId)}`,
      { auth: "apiKey", query: { period } },
    );
  }

  async listLeaderboardTop(
    boardId: string,
    params?: { period?: string; page_size?: number; page_token?: string },
  ): Promise<ListLeaderboardTopResponse> {
    return this.http.request<ListLeaderboardTopResponse>(
      "GET",
      `/v1/server/leaderboards/${encodeURIComponent(boardId)}/top`,
      { auth: "apiKey", query: listQuery(params as never) },
    );
  }

  /** 结算读（Phase 2）：发奖明细 = 用户资产 ledger 的 leaderboard_settlement 引用视图。 */
  async getLeaderboardSettlement(
    boardId: string,
    period: string,
  ): Promise<{
    settlement: LeaderboardSettlement;
    grants: LeaderboardSettlementGrant[];
  }> {
    return this.http.request(
      "GET",
      `/v1/server/leaderboards/${encodeURIComponent(boardId)}/settlements/${encodeURIComponent(period)}`,
      { auth: "apiKey" },
    );
  }

  async listLeaderboardSettlements(
    boardId: string,
    limit?: number,
  ): Promise<{ settlements: LeaderboardSettlement[] }> {
    return this.http.request(
      "GET",
      `/v1/server/leaderboards/${encodeURIComponent(boardId)}/settlements`,
      { auth: "apiKey", query: { limit } },
    );
  }

  // —— board 配置管控（预置与门禁自动化；Create/Update 需 leaderboards.admin）——

  /** 幂等建榜：已存在且配置逐字段相等（缺省归一后）→ 200 + 现状（重放安全）；
   * 不等 → 409 ALREADY_EXISTS（错误消息附字段 diff）。 */
  async createLeaderboardBoard(
    id: string,
    input: CreateLeaderboardBoardInput,
  ): Promise<LeaderboardBoard> {
    return this.http.request<LeaderboardBoard>(
      "POST",
      "/v1/server/leaderboards/boards",
      { auth: "apiKey", body: { id, ...input } },
    );
  }

  /** 榜配置读（门禁断言主读点：period/tz/policy/value bounds/client_submit 等全文）。 */
  async getLeaderboardBoard(boardId: string): Promise<LeaderboardBoard> {
    return this.http.request<LeaderboardBoard>(
      "GET",
      `/v1/server/leaderboards/boards/${encodeURIComponent(boardId)}`,
      { auth: "apiKey" },
    );
  }

  async listLeaderboardBoards(): Promise<{ boards: LeaderboardBoard[] }> {
    return this.http.request("GET", "/v1/server/leaderboards/boards", { auth: "apiKey" });
  }

  /** 已有条目的期 key 降序（对账期存在性）。 */
  async listLeaderboardBoardPeriods(
    boardId: string,
    limit?: number,
  ): Promise<{ periods: string[] }> {
    return this.http.request(
      "GET",
      `/v1/server/leaderboards/boards/${encodeURIComponent(boardId)}/periods`,
      { auth: "apiKey", query: { limit } },
    );
  }

  /** 改榜（optional = 不修改；显式清空走 clear_*）。不含 rewards——奖励规则
   * 编辑仅 console owner。 */
  async updateLeaderboardBoard(
    boardId: string,
    input: UpdateLeaderboardBoardInput,
  ): Promise<LeaderboardBoard> {
    return this.http.request<LeaderboardBoard>(
      "PATCH",
      `/v1/server/leaderboards/boards/${encodeURIComponent(boardId)}`,
      { auth: "apiKey", body: input },
    );
  }
}
